crypto/bio/b_sock.c

```diff
@@ -369,4 +375,110 @@ int BIO_sock_info(int sock,
     return 1;
 }
 
-#endif
+/* TODO simplify by BIO_socket_wait() further other uses of select() in apps/ */
+/*
+ * Wait on fd at most until max_time; succeed immediately if max_time == 0.
+ * If for_read == 0 then assume to wait for writing, else wait for reading.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+int BIO_socket_wait(int fd, int for_read, time_t max_time)
+{
+    fd_set confds;
+    struct timeval tv;
+    time_t now;
+
+    if (max_time == 0)
+        return 1;
+
+    now = time(NULL);
+    if (max_time <= now)
+        return 0;
+
+    FD_ZERO(&confds);
+    openssl_fdset(fd, &confds);
+    tv.tv_usec = 0;
+    tv.tv_sec = (long)(max_time - now); /* this might overflow */
+    return select(fd + 1, for_read ? &confds : NULL,
+                  for_read ? NULL : &confds, NULL, &tv);
+}
+
+/*
+ * Wait on BIO at most until max_time; succeed immediately if max_time == 0.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+static int bio_wait(BIO *bio, time_t max_time)
+{
+    int fd;
+
+    if (BIO_get_fd(bio, &fd) <= 0)
+        return -1;
+    return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
+}
+
+/*
+ * Wait on BIO at most until max_time; succeed immediately if max_time == 0.
+ * Call BIOerr(...) unless success.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+int BIO_wait(BIO *bio, time_t max_time)
+{
+    int rv = bio_wait(bio, max_time);
+
+    if (rv <= 0)
+        BIOerr(0, rv == 0 ? BIO_R_TRANSFER_TIMEOUT : BIO_R_TRANSFER_ERROR);
+    return rv;
+}
+
+/*
+ * Connect via the given BIO using BIO_do_connect() until success/timeout/error.
+ * Parameter timeout == 0 means infinite, < 0 leads to immediate timeout error.
+ * Returns -1 on error, 0 on timeout, and 1 on success.
+ */
+int BIO_connect_retry(BIO *bio, int timeout)
```

Comment:
So now the BIO might have been passed in from the application...I wonder how this function copes with something like a mem bio?

Reply:
Good thought!

I've meanwhile tried with a mem BIO, and it turns out that the function fails (as expected) but without an entry in the error queue. It turns out that a
```
        BIOerr(0, BIO_R_UNSUPPORTED_METHOD);
```
was missing in the default case of `mem_ctrl()` in `crypto/bio/bss_mem.c`.

I've simply added this line (as a separate commit in this PR) rather than issuing an independent PR.
Moreover, since it may happen that the BIO call fails also for other reasons (and maybe for a different BIO type) without adding an error queue entry I've added a catch-up to `BIO_connect_retry()`.

Doing this I found that also the distinction within `BIO_connect_retry()` between timeout and any other type of error could be improved, which I did.

Reply:
My goodness, adding the error queue entry lead to several SSL test failures.
There must be something pretty fishy in the SSL state machine.
I'll transform this to an independent issue.
So for the mem BIO not giving a suitable error on unsupported `BIO_C_DO_STATE_MACHINE` the workaround I just introduced in `BIO_connect_retry()` applies.

Reply:
So, I'm a little unclear, is the result of this issue that we require BIOs to be capable of handling `BIO_C_DO_STATE_MACHINE` for us to be able to use the HTTP client with them?

Reply:
An HTTP client needs a BIO that can be used for connecting over a network,
and AFAICS this boils down to supporting at least `BIO_C_DO_STATE_MACHINE`.

If this is not the case the HTTP functions will yield `BIO_R_CONNECT_ERROR` (or any more specific error message like `BIO_R_UNSUPPORTED_METHOD` if the respective BIO implementation gives such error detail, which currently is not the case for mem BIOs due to #10955).

Reply:
Shall I make explicit in the documentation that if a connection BIO is provided to the HTTP client it naturally needs to support a networked connection?

Reply:
I'm wondering  if there is anything we can do to just ignore the error if the `BIO_connnect_retry()` call fails? Failing that I guess we'll have to document the restriction

Reply:
I do not understand what you mean here. 
We do not and cannot ignore any error returned by `BIO_connect_retry()` - or why/when do you think this would make sense?
We usually cannot ignore errors that the called function `BIO_do_connect()` (and thus errors by the underlying BIO) may return, except where a retry is due for some reason. 

The issue mentioned above regarding non-supported `BIO_C_DO_STATE_MACHINE` only affects the exact error detail provided: ideally the underlying BIO provides a to-the-point error entry such as `BIO_R_UNSUPPORTED_METHOD`, but if it returns `0` (failure) without pushing such an entry the enhancement I added to `BIO_connect_retry()` pushes the more general entry `BIO_R_CONNECT_ERROR` as a workaround. In any case it propagates the negative outcome of  `BIO_do_connect()`.

Reply:
Well, fundamentally I see no reason why someone should not be able to use a mem bio (or any other "non network" BIO) to do HTTP. For example, often people will use a mem bio as an intermediate step before pushing on to some underlying network connection that is unknown to OpenSSL. I'm just trying to see if there is a way to not force restrictions on what type of BIO can be used. I don't think we do that anywhere else and it detracts significantly from the power of a BIO.

Reply:
Ah, now I understand!
Can the same (mem or other non-network) BIO be used at first for writing (HTTP request) data to and then for reading (HTTP response) data from?

Reply:
Hmmm. Well actually you often have a different BIO for reading and writing. The sslapitest does this with clients and servers communicating entirely via mem bios. There is one mem BIO for the "write" side on the client - and that is used as the "read" side on the server. Similarly there is another mem BIO on the "read" side for the client and that is used as the "write" side on the server.

Reply:
You can also use a "bio pair" which probably enables you to do this with a single BIO.

Reply:
This makes more sense to me.
To support fully flexible BIO-based HTTP, I can add optional `wbio` and `rbio` parameters.
Shall I also keep the recently added `cbio` parameter, which may be used to provide a custom connection BIO which is then also used for reading and writing data (such that  `wbio` and `rbio` are ignored in this case)?

Reply:
My last response overlapped with your most recent one.

Reply:
Yes, a BIO pair sounds even better.

Reply:
Hmm, I've meanwhile had a look at the documentation of BIO pairs: https://www.openssl.org/docs/manmaster/man3/BIO_new_bio_pair.html which I find hard to understand.
I've also searched for uses of BIO pairs within OpenSSL and found just one, namely in an apparently outdated test: `test/ssltest_old.c`.
So I think it's better to keep away from that and instead follow the model you mentioned before, namely to allow to give two BIOs: one for writing requests and one to read responses from.

Reply:
I've replaced the optional `cbio` parameter by the more flexible optional `wbio` and `rbio` parameters, which was rather straightforward, and updated the docs accordingly.

This was the perfect moment to add also an HTTP client unit test since now mem BIOs can be used to pseudo-connect to an internal mock HTTP server. The tests cover nearly all client code, including GET redirections. This way I discovered and fixed an issue with server-relative redirection locations.